Add ensemble scheduler

GuanLuo opened the pull request.

deadeyegoodwin commented on src/core/backend.cc:
```diff
@@ -241,9 +242,11 @@ InferenceBackend::SetConfiguredScheduler(
     RETURN_IF_ERROR(SequenceBatchScheduler::Create(
         config_, runner_cnt, OnInit, OnRun, &scheduler));
-  } else {
+  } else if (config_.has_dynamic_batching()) {
```
Dynamic batching has to be the default choice because if the model configuration does not list any scheduler then it should use the dynamic scheduler.

deadeyegoodwin commented on src/core/backend.h:
```diff
@@ -102,6 +102,9 @@ class InferenceBackend {
       Scheduler::StandardRunFunc OnRun);
 
+  // Get the raw pointer to the scheduler of the servable.
+  Scheduler* GetRawScheduler() { return scheduler_.get(); }
```
Just "Scheduler". I know there are some Get prefixes on other functions but I plan to fix those.

deadeyegoodwin commented on src/servables/ensemble/ensemble_bundle.cc:
```diff
@@ -57,10 +57,9 @@ EnsembleBundle::Init(const std::string& path, const ModelConfig& config)
+  EnsembleScheduler* scheduler =
+      static_cast<EnsembleScheduler*>(GetRawScheduler());
+  RETURN_IF_ERROR(scheduler->SetInferenceServer(inference_server));
   return InferenceBackend::SetInferenceServer(inference_server);
```
It seems like you don't actually have to store inference_server_ in backend.h. The SetInferencServer() method can just do nothing by default.

deadeyegoodwin commented on src/core/ensemble_scheduler.h:
```diff
@@ -0,0 +1,70 @@
+      std::function<void(Status)> OnComplete) override;
+
+  // Set the inference server that the scheduler is communicating with
+  Status SetInferenceServer(void* inference_server)
```
Override

deadeyegoodwin replied:
Override is still missing on SetInferenceServer

GuanLuo replied:
This is a new function for EnsembleScheduler, the Scheduler interface doesn't have this function. Should I also add this to Scheduler and default to do nothing?

deadeyegoodwin replied:
My mistake, I thought it was the override from Backend.

deadeyegoodwin commented on src/core/ensemble_scheduler.cc:
```diff
@@ -0,0 +1,245 @@
+
+  // Helper function that returns a list of step that should be run
+  // under current state.
+  // 'ok' will be set to false if the ensemble context will not make any
```
'ok' is not a good name. Do you 'deadlocked'? Or 'complete'?

deadeyegoodwin commented on src/core/ensemble_scheduler.cc:
```diff
@@ -0,0 +1,245 @@
+    : is_(is), config_(config), ensemble_graph_(ensemble_graph), stats_(stats),
+      request_provider_(request_provider),
+      response_provider_(response_provider), OnComplete_(OnComplete)
+{
```
At some point, perhaps here, you need to grab servable handles (backends) for all the models you will need in the ensemble. Otherwise they could be unloaded while the ensemble is executing.

deadeyegoodwin commented on src/core/ensemble_scheduler.cc:
```diff
@@ -0,0 +1,531 @@
+Status
+EnsembleContext::FinishEnsemble()
+{
+  if (stats_ == nullptr) {
```
Overloading stats for this purpose is confusing...

deadeyegoodwin commented on src/core/ensemble_scheduler.cc:
```diff
@@ -0,0 +1,531 @@
+      }
+    }
+    // Error or no more progress (completed or deadlock)
+    if ((!ensemble_status_.IsOk()) || (inflight_request_counter_ == 0)) {
```
inflight_steps_counter_ would be a more meaningful name. I assume this is just to be extra safe since you verify that the emsemble graph is fully connected and has no loops?

Also, shouldn't you set some kind of error if you do finish the ensemble because of a deadlock?

deadeyegoodwin commented on src/core/ensemble_scheduler.cc:
```diff
@@ -0,0 +1,531 @@
+  steps.clear();
+
+  std::set<size_t> next_step_idx;
+  // Get steps whose tensors used for input are set
```
This is not a very efficient way to execute the graph. We could have a more direct represntation of the graph so we don't have to iterate over all the tensors. It is fine for a first implementation but at some point we may need to refine it.

GuanLuo commented:
New changes since last review starts from this commit https://github.com/NVIDIA/tensorrt-inference-server/pull/154/commits/fc784b60833bb698ef4fc52fe43063d0d79ae482
